README.md

## Jalali/Shamsi Calendar  Datepicker and TimePicker
Calendar , Datepicker and Date Convertor For Persian Date(Jalali/Shamsi date) with holiday for Android and Ios

Based on Material  Flutter DatePicker And TimePicker


![](https://github.com/mrmostafaei/jalalidatepicker/blob/main/screenshots/screenshots.gif)

## Usage

## Import this line in Flutter pubspec
```dart
jalalidatepicker: <Last Version>
```
## To  Use Calendar Or DatePicker  , Import this line to your dart file
```dart
import 'package:jalalidatepicker/jalalidatepicker.dart';
```
## To  Use date Convertor , Import this line to your dart file
```dart
import 'package:persian_date/persian_date.dart';
```

After import library,  to show the calendar you can use jalaliCalendarPicker() constructor method , on the constructor `context` is important
if you want to return the selected date to Gregorian, on the constructor set the convertToGregorian to true

## This sample return the selected date as Jalali(شمسی)
```dart
Future _selectDate() async {
    String picked = await jalaliCalendarPicker(context: context); // نمایش خروجی به صورت شمسی
   //  await jalaliCalendarPicker(context: context,convertToGregorian: true); // نمایش خروجی به صورت میلادی
    if (picked != null) setState(() => _value =picked.toString());
    print(_value);
  }
```

## This sample return the selected date as Gregorian(میلادی)
```dart
Future _selectDate() async {
    String picked = await jalaliCalendarPicker(context: context,convertToGregorian: true); // نمایش خروجی به صورت میلادی
    if (picked != null) setState(() => _value =picked.toString());
    print(_value);
  }
```

## Those parameters using  To Show the TimePicker

|   Parameter    |            operation                           |
|:--------------:|:----------------------------------------------:|
| showTimePicker | To show TimePicker set it true                 |
| hore24Format   | To return time in 24 hours format set it true  |
| initialTime    | To set initial time use it with TimeOfDay      |



## To use datePicker you can follow this sample


To show DatePicker those parameters is required
```dart
minYear // min date this required  for the start year in Jalali
maxYear // max  date this required  for the last year in jalali
confirm and cancel // widget for show text
_format // for returning date like String _format = 'yyyy-mm-dd';
onChanged // return date when the user changes the date
onConfirm //  return date when the user click confirm
```

For set, default date use those parameters
```dart
initialYear => for year
initialMonth => for month
initialDay => for day
```
## Full datePicker Sample
```dart
    DatePicker.showDatePicker(
      context,
      minYear: 1300,
      maxYear: 1450,
      confirm: Text(
        'Confirm',
        style: TextStyle(color: Colors.red),
      ),
      cancel: Text(
        'Cancel',
        style: TextStyle(color: Colors.cyan),
      ),
      dateFormat: _format,
      onChanged: (year, month, day) {
        if (!showTitleActions) {
          _changeDatetime(year, month, day);
        }
      },
      onConfirm: (year, month, day) {
        _changeDatetime(year, month, day);
        _valuePiker = " Full date : $_datetime  \n year : $year \n  Month  :   $month \n  day :  $day";
      }
    )

 void _changeDatetime(int year, int month, int day) {
    setState(() {
      _datetime = '$year-$month-$day';
    });
  }

```

## Full Date Convertor Sample
```dart
PersianDate pDate = PersianDate(gregorian: "1989-01-29");
    print("Now ${pDate.getDate}");

    PersianDate persianDate = PersianDate();
    print("Now ${persianDate.now}");
    print(persianDate.hour);
    print("year ${persianDate.year}");
    print("isHoliday ${persianDate.isHoliday}");
    print("isHoliday ${persianDate.weekdayname}");
    print(persianDate.monthname); // نام ماه
      print(persianDate.month); // ماه
      print(persianDate.day); // روز
      print(persianDate.hour);// ساعت
      print(persianDate.minute);// دقیقه
      print(persianDate.second);// ثانیه
      print(persianDate.millisecond); // میلی ثانیه
      print(persianDate.microsecond);//


print("Parse Gregorian To Jalali ${persianDate.gregorianToJalali("2019-02-20T00:19:54.000Z","yyyy-m-d hh:nn")}");
print("Parse Jalali To Gregorian ${persianDate.jalaliToGregorian("1368-05-30 19:54", "yyyy-m-d hh:nn")}");
```

## those formats are supported in date converter
```dart
   `"2012-02-27 13:27:00"`
    `"2012-02-27 13:27:00.123456z"`
    `"2012-02-27 13:27:00,123456z"`
    `"20120227 13:27:00"`
    `"20120227T132700"`
    `"20120227"`
    `"+20120227"`
    `"2012-02-27T14Z"`
    `"2012-02-27T14+00:00"`
    `"-123450101 00:00:00 Z"`: in the year -12345.
    `"2002-02-27T14:00:00-0500"`: Same as `"2002-02-27T19:00:00Z"`
```


| Format |            operation            | Format |            operation            |
|:------:|:-------------------------------:|:------:|:-------------------------------:|
| yyyy   | return year with Four number    | hh     | hours with  two number          |
| yy     | return year with two number     | h      | hours with one number           |
| mm     | return month with two number    | HH     | hours with  two number in 24    |
| m      | return month with one number    | H      | hours with one number in 24     |
| MM     | the month name with full letter | nn     | minut with two number           |
| M      | the month name with short form  | n      | minut with one number           |
| dd     | the day with two number         | ss     | Second with two number          |
| d      | the day with one number         | s      | Second with one number          |
| w      | week number                     | SS     | show milliSecond                |
| DD     | the day name with full letter   | S      | show milliSecond                |
| D      | the day name with  short form   | uuu    | show microsecond                |
| am     | Show time in short              | u      | show microsecond                |
| AM     | Show full time                  |   .    |          .                      |


## bug report
Contact me if you have any problems with this library or have any suggestion [tracker]:info@milad-tech.com


CHANGELOG.md
## 1.0.4
* Upgrade dart 3
* fixed bugs
* format code

## 1.0.3+1
* format code

## 1.0.3
* Fixed displayed month bug

## 1.0.2
* delete print method from lib

## 1.0.1
* Update Some code

## 1.0.0
* Fixed Selected Day
* Relpace intl package with jiffy
* Migrate to Androidx

## 0.1.6+2
* Fixed Bug

## 0.1.6+1
* Fixed TimePicker

## 0.1.6
* Add TimePicker

## 0.1.5
* Update Some Code And ReadMe File 

## 0.1.4
* Update Some Code 

## 0.1.3
* Update and fixed Some Code

## 0.1.2
* Update and fixed Some Code 
* returned date in Gregorian 
* fixed bug on some devices

## 0.1.1
* fixed Bug  

## 0.1.0
* fixed Bug 

## 0.0.2
* fixed Bug 

## 0.0.1
* TODO: Describe initial release.


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:jalalidatepicker/src/datepicker.dart';
import 'package:jalalidatepicker/src/jalalicalendarpicker.dart';
import 'package:jalalidatepicker/src/persian_date.dart';

void main() {
  runApp(new MaterialApp(
    home: new MyApp(),
  ));
}

class MyApp extends StatefulWidget {
  @override
  _State createState() => new _State();
}

class _State extends State<MyApp> {
  PersianDate persianDate = PersianDate("yyyy/mm/dd  \n DD  , d  MM  ");
  String _datetime = '';
  String _format = 'yyyy-mm-dd';
  String _value = '';
  String _valuePiker = '';
  DateTime selectedDate = DateTime.now();

  Future _selectDate() async {
    String? picked = await jalaliCalendarPicker(
        context: context,
        convertToGregorian: false,
        showTimePicker: true,
        hore24Format: true);
    if (picked != null) setState(() => _value = picked);
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text('Persain Date'),
        centerTitle: true,
      ),
      body: new Container(
        child: Column(
          children: <Widget>[
            Center(
              child: Column(
                children: <Widget>[
                  Text('  مبدّل تاریخ و زمان ,‌ تاریخ هجری شمسی '),
                  Text(' تقویم شمسی '),
                  Text('date picker شمسی '),
                  new ElevatedButton(
                    onPressed: _selectDate,
                    child: new Text('نمایش تقویم'),
                  ),
                  new ElevatedButton(
                    onPressed: _showDatePicker,
                    child: new Text('نمایش دیت پیکر'),
                  ),
                  Text(
                    "\nزمان و تاریخ فعلی سیستم :  ${persianDate.getNow}",
                    textAlign: TextAlign.center,
                    textDirection: TextDirection.rtl,
                  ),
                  Divider(),
                  Text(
                    "تقویم ",
                    textAlign: TextAlign.center,
                  ),
                  Text(
                    _value,
                    textAlign: TextAlign.center,
                  ),
                  Divider(),
                  Text(
                    _valuePiker,
                    textAlign: TextAlign.center,
                  ),
                ],
              ),
            )
            // Expanded(child: ShowCalender())
          ],
        ),
      ),
    );
  }

  /// Display date picker.
  void _showDatePicker() {
    final bool showTitleActions = false;
    DatePicker.showDatePicker(context,
        minYear: 1300,
        maxYear: 1450,
/*      initialYear: 1368,
      initialMonth: 05,
      initialDay: 30,*/
        confirm: Text(
          'تایید',
          style: TextStyle(color: Colors.red),
        ),
        cancel: Text(
          'لغو',
          style: TextStyle(color: Colors.cyan),
        ),
        dateFormat: _format, onChanged: (year, month, day) {
      if (!showTitleActions) {
        _changeDatetime(year!, month!, day!);
      }
    }, onConfirm: (year, month, day) {
      _changeDatetime(year!, month!, day!);
      _valuePiker =
          " تاریخ ترکیبی : $_datetime  \n سال : $year \n  ماه :   $month \n  روز :  $day";
    });
  }

  void _changeDatetime(int year, int month, int day) {
    setState(() {
      _datetime = '$year-$month-$day';
    });
  }
}


android/src/main/java/com/miladmostafaei/jalali/calendar/jalalicalendar/JalaliCalendarPlugin.java
package com.miladmostafaei.jalali.calendar.jalalicalendar;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** JalaliCalendarPlugin */
public class JalaliCalendarPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "jalalidatepicker");
    channel.setMethodCallHandler(new JalaliCalendarPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }
}


lib/src/datepicker.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'persian_date.dart';
typedef DateChangedCallback(int? year, int? month, int? date);

const String _kDateFormatDefault = 'yyyy-mm-dd';

const double _kDatePickerHeight = 210.0;
const double _kDatePickerTitleHeight = 44.0;
const double _kDatePickerItemHeight = 36.0;
const double _kDatePickerFontSize = 18.0;

const int _kDefaultMinYear = 1300;
const int _kDefaultMaxYear = 1450;

const int _kMonthCount = 12;

const List<int> leapYearMonths = const <int>[1, 3, 5, 7, 8, 10, 12];

class DatePicker {
  static void showDatePicker(
    BuildContext context, {
    bool showTitleActions = true,
    int minYear = _kDefaultMinYear,
    int maxYear = _kDefaultMaxYear,
    int? initialYear,
    int? initialMonth,
    int? initialDay,
    Widget? cancel,
    Widget? confirm,
    DateChangedCallback? onChanged,
    DateChangedCallback? onConfirm,
    dateFormat = _kDateFormatDefault,
  }) {
    if (dateFormat == null || dateFormat.length == 0) {
      dateFormat = _kDateFormatDefault;
    }

    PersianDate now = PersianDate();
    if (initialYear == null) {
      initialYear = now.year;
    }
    if (initialMonth == null) {
      initialMonth = now.month;
    }
    if (initialDay == null) {
      initialDay = now.day;
    }

    Navigator.push(
      context,
      new _DatePickerRoute(
        showTitleActions: showTitleActions,
        minYear: minYear,
        maxYear: maxYear,
        initialYear: initialYear,
        initialMonth: initialMonth,
        initialDate: initialDay,
        cancel: cancel,
        confirm: confirm,
        onChanged: onChanged,
        onConfirm: onConfirm,
        dateFormat: dateFormat,
        theme: Theme.of(context),
        barrierLabel:
            MaterialLocalizations.of(context).modalBarrierDismissLabel,
      ),
    );
  }
}

class _DatePickerRoute<T> extends PopupRoute<T> {
  _DatePickerRoute({
    this.showTitleActions,
    this.minYear,
    this.maxYear,
    this.initialYear,
    this.initialMonth,
    this.initialDate,
    this.cancel,
    this.confirm,
    this.onChanged,
    this.onConfirm,
    this.theme,
    this.barrierLabel,
    this.locale,
    this.dateFormat,
    RouteSettings? settings,
  }) : super(settings: settings);

  final bool? showTitleActions;
  final int? minYear, maxYear, initialYear, initialMonth, initialDate;
  final Widget? cancel, confirm;
  final DateChangedCallback? onChanged;
  final DateChangedCallback? onConfirm;
  final ThemeData? theme;
  final String? locale;
  final String? dateFormat;

  @override
  Duration get transitionDuration => const Duration(milliseconds: 200);

  @override
  bool get barrierDismissible => true;

  @override
  final String? barrierLabel;

  @override
  Color get barrierColor => Colors.black54;

  AnimationController? _animationController;

  @override
  AnimationController createAnimationController() {
    assert(_animationController == null);
    _animationController =
        BottomSheet.createAnimationController(navigator!.overlay!);
    return _animationController!;
  }

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    Widget bottomSheet = new MediaQuery.removePadding(
      context: context,
      removeTop: true,
      child: _DatePickerComponent(
        minYear: minYear,
        maxYear: maxYear,
        initialYear: initialYear,
        initialMonth: initialMonth,
        initialDate: initialDate,
        cancel: cancel,
        confirm: confirm,
        onChanged: onChanged,
        locale: locale,
        dateFormat: dateFormat,
        route: this,
      ),
    );
    if (theme != null) {
      bottomSheet = new Theme(data: theme!, child: bottomSheet);
    }
    return bottomSheet;
  }
}

class _DatePickerComponent extends StatefulWidget {
  _DatePickerComponent(
      {
      required this.route,
      this.minYear = _kDefaultMinYear,
      this.maxYear = _kDefaultMaxYear,
      this.initialYear = -1,
      this.initialMonth = 1,
      this.initialDate = 1,
      this.cancel,
      this.confirm,
      this.onChanged,
      this.locale,
      this.dateFormat});

  final DateChangedCallback? onChanged;
  final int? minYear, maxYear, initialYear, initialMonth, initialDate;

  final Widget? cancel;
  final Widget? confirm;

  final _DatePickerRoute route;

  final String? locale;
  final String? dateFormat;

  @override
  State<StatefulWidget> createState() => _DatePickerState(this.minYear,
      this.maxYear, this.initialYear, this.initialMonth, this.initialDate);
}

class _DatePickerState extends State<_DatePickerComponent> {
  final int? minYear, maxYear;
  int? _currentYear, _currentMonth, _currentDate;
  int? _dateCountOfMonth;
  FixedExtentScrollController? yearScrollCtrl, monthScrollCtrl, dateScrollCtrl;

  _DatePickerState(this.minYear, this.maxYear, this._currentYear,
      this._currentMonth, this._currentDate) {
    if (this._currentYear == -1) {
      this._currentYear = this.minYear;
    }
    if (this._currentYear! < this.minYear!) {
      this._currentYear = this.minYear;
    }
    if (this._currentYear! > this.maxYear!) {
      this._currentYear = this.maxYear;
    }

    if (this._currentMonth! < 1) {
      this._currentMonth = 1;
    }
    if (this._currentMonth! > 12) {
      this._currentMonth = 12;
    }

    if (this._currentDate! < 1) {
      this._currentDate = 1;
    }
    if (this._currentDate! > 31) {
      this._currentDate = 31;
    }

    yearScrollCtrl = new FixedExtentScrollController(
        initialItem: _currentYear! - this.minYear!);
    monthScrollCtrl =
        new FixedExtentScrollController(initialItem: _currentMonth! - 1);
    dateScrollCtrl =
        new FixedExtentScrollController(initialItem: _currentDate! - 1);
    _dateCountOfMonth = _calcDateCount();
  }

  @override
  Widget build(BuildContext context) {
    return new GestureDetector(
      child: new AnimatedBuilder(
        animation: widget.route.animation!,
        builder: (BuildContext context, Widget? child) {
          return new ClipRect(
            child: new CustomSingleChildLayout(
              delegate: new _BottomPickerLayout(widget.route.animation!.value,
                  showTitleActions: widget.route.showTitleActions),
              child: new GestureDetector(
                child: Material(
                  color: Colors.transparent,
                  child: _renderPickerView(),
                ),
              ),
            ),
          );
        },
      ),
    );
  }

  void _setYear(int index) {
    int year = widget.minYear! + index;
    if (_currentYear != year) {
      _currentYear = year;
      _notifyDateChanged();
    }
  }

  void _setMonth(int index) {
    int month = index + 1;
    if (_currentMonth != month) {
      _currentMonth = month;
      int dateCount = _calcDateCount();
      if (_dateCountOfMonth != dateCount) {
        setState(() {
          _dateCountOfMonth = dateCount;
        });
      }
      if (_currentDate! > dateCount) {
        _currentDate = dateCount;
      }
      _notifyDateChanged();
    }
  }

  void _setDate(int index) {
    int date = index + 1;
    if (_currentDate != date) {
      _currentDate = date;
      _notifyDateChanged();
    }
  }

  static const List<int> _daysInMonth = <int>[
    31,
    31,
    31,
    31,
    31,
    31,
    30,
    30,
    30,
    30,
    30,
    -1
  ];
  static const List<int> _kabise = <int>[1, 5, 9, 13, 17, 22, 26, 30];

  int _calcDateCount() {
    if (_currentMonth == 12) {
      var modeyear = _currentYear! % 33;
      if (_kabise.indexOf(modeyear) != -1) {
        return 30;
      }
      return 29;
    }
    return _daysInMonth[_currentMonth! - 1];
  }

  void _notifyDateChanged() {
    if (widget.onChanged != null) {
      widget.onChanged!(_currentYear, _currentMonth, _currentDate);
    }
  }

  Widget _renderPickerView() {
    Widget itemView = _renderItemView();
    if (widget.route.showTitleActions!) {
      return Column(
        children: <Widget>[
          _renderTitleActionsView(),
          itemView,
        ],
      );
    }
    return itemView;
  }

  Widget _renderYearsPickerComponent(String yearAppend) {
    return new Expanded(
      flex: 1,
      child: Container(
        padding: EdgeInsets.all(8.0),
        height: _kDatePickerHeight,
        decoration: BoxDecoration(color: Colors.white),
        child: CupertinoPicker(
          backgroundColor: Colors.white,
          scrollController: yearScrollCtrl,
          itemExtent: _kDatePickerItemHeight,
          onSelectedItemChanged: (int index) {
            _setYear(index);
          },
          children:
              List.generate(widget.maxYear! - widget.minYear! + 1, (int index) {
            return Container(
              height: _kDatePickerItemHeight,
              alignment: Alignment.center,
              child: Text(
                '${widget.minYear! + index}$yearAppend',
                style: TextStyle(
                    color: Color(0xFF000046), fontSize: _kDatePickerFontSize),
                textAlign: TextAlign.start,
              ),
            );
          }),
        ),
      ),
    );
  }

  Widget _renderMonthsPickerComponent(String monthAppend, {String? format}) {
    return new Expanded(
      flex: 1,
      child: Container(
          padding: EdgeInsets.all(8.0),
          height: _kDatePickerHeight,
          decoration: BoxDecoration(color: Colors.white),
          child: CupertinoPicker(
            backgroundColor: Colors.white,
            scrollController: monthScrollCtrl,
            itemExtent: _kDatePickerItemHeight,
            onSelectedItemChanged: (int index) {
              _setMonth(index);
            },
            children: List.generate(_kMonthCount, (int index) {
              return Container(
                height: _kDatePickerItemHeight,
                alignment: Alignment.center,
                child: Row(
                  children: <Widget>[
                    new Expanded(
                        child: Text(
                      (format == null)
                          // index is 0,1,2...11  month is 1,2,3...12
                          ? '${index + 1}$monthAppend'
                          : '${_formatMonthComplex(index, format)}$monthAppend',
                      style: TextStyle(
                          color: Color(0xFF000046),
                          fontSize: _kDatePickerFontSize),
                      textAlign: TextAlign.center,
                      softWrap: false,
                      overflow: TextOverflow.fade,
                    ))
                  ],
                ),
              );
            }),
          )),
    );
  }

  Widget _renderDaysPickerComponent(String dayAppend) {
    return new Expanded(
      flex: 1,
      child: Container(
          padding: EdgeInsets.all(8.0),
          height: _kDatePickerHeight,
          decoration: BoxDecoration(color: Colors.white),
          child: CupertinoPicker(
            backgroundColor: Colors.white,
            scrollController: dateScrollCtrl,
            itemExtent: _kDatePickerItemHeight,
            onSelectedItemChanged: (int index) {
              _setDate(index);
            },
            children: List.generate(_dateCountOfMonth!, (int index) {
              return Container(
                height: _kDatePickerItemHeight,
                alignment: Alignment.center,
                child: Text(
                  "${index + 1}$dayAppend",
                  style: TextStyle(
                      color: Color(0xFF000046), fontSize: _kDatePickerFontSize),
                  textAlign: TextAlign.start,
                ),
              );
            }),
          )),
    );
  }

  Widget _renderItemView() {
    String yearAppend = "";
    String monthAppend = "";
    String dayAppend = "";

    List<Widget> pickers =  [];
    List<String> formatSplit = widget.dateFormat!.split('-');
    for (int i = 0; i < formatSplit.length; i++) {
      var format = formatSplit[i];
      if (format.contains("yy")) {
        pickers.add(_renderYearsPickerComponent(yearAppend));
      } else if (format.contains("mm")) {
        pickers.add(_renderMonthsPickerComponent(monthAppend, format: format));
      } else if (format.contains("dd")) {
        pickers.add(_renderDaysPickerComponent(dayAppend));
      }
    }

    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: pickers,
    );
  }

  String _digits(int? value, int length) {
    String ret = '$value';
    if (ret.length < length) {
      ret = '0' * (length - ret.length) + ret;
    }
    return ret;
  }

  // Title View
  Widget _renderTitleActionsView() {
    String done = "تایید";
    String cancel = "لغو";

    Widget? cancelWidget = this.widget.cancel;
    if (cancelWidget == null) {
      cancelWidget = Text(
        '$cancel',
        style: TextStyle(
          color: Theme.of(context).unselectedWidgetColor,
          fontSize: 16.0,
        ),
      );
    }

    Widget? confirmWidget = this.widget.confirm;
    if (confirmWidget == null) {
      confirmWidget = Text(
        '$done',
        style: TextStyle(
          color: Theme.of(context).primaryColor,
          fontSize: 16.0,
        ),
      );
    }

    return Container(
      height: _kDatePickerTitleHeight,
      decoration: BoxDecoration(color: Colors.white),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Container(
            height: _kDatePickerTitleHeight,
            child: TextButton(
              child: cancelWidget,
              onPressed: () => Navigator.pop(context),
            ),
          ),
          Container(
            height: _kDatePickerTitleHeight,
            child: TextButton(
              child: confirmWidget,
              onPressed: () {
                if (widget.route.onConfirm != null) {
                  widget.route.onConfirm!(
                      int.parse(_digits(_currentYear, 2)),
                      int.parse(_digits(_currentMonth, 2)),
                      int.parse(_digits(_currentDate, 2)));
                }
                Navigator.pop(context);
              },
            ),
          ),
        ],
      ),
    );
  }

  // format month
  String _formatMonthComplex(int month, String format) {
    if (widget.locale == null) {
      return (month + 1).toString();
    }
    return (month + 1).toString();
  }
}

class _BottomPickerLayout extends SingleChildLayoutDelegate {

  final double progress;
  final bool? showTitleActions;

  _BottomPickerLayout(this.progress, { this.showTitleActions});

  @override
  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
    double maxHeight = _kDatePickerHeight;
    if (showTitleActions!) {
      maxHeight += _kDatePickerTitleHeight;
    }

    return new BoxConstraints(
        minWidth: constraints.maxWidth,
        maxWidth: constraints.maxWidth,
        minHeight: 0.0,
        maxHeight: maxHeight);
  }

  @override
  Offset getPositionForChild(Size size, Size childSize) {
    double height = size.height - childSize.height * progress;
    return new Offset(0.0, height);
  }

  @override
  bool shouldRelayout(_BottomPickerLayout oldDelegate) {
    return progress != oldDelegate.progress;
  }
}
